vm/devices/storage/storvsc_driver/src/lib.rs

```diff
@@ -0,0 +1,856 @@
+}
+
+struct PendingOperation {
+    sender: Sender<StorvscCompletion>,
+}
+
+impl PendingOperation {
+    fn new(sender: Sender<StorvscCompletion>) -> Self {
+        Self { sender }
+    }
+
+    fn complete(&mut self, result: storvsp_protocol::ScsiRequest) -> Result<(), StorvscError> {
+        self.sender
+            .send_blocking(StorvscCompletion {
+                completion: Some(result),
+            })
+            .map_err(|_err| StorvscError::NotificationError)
+    }
+
+    fn cancel(&mut self) {
+        // Sending completion with an empty result indicates cancellation or other error.
+        self.sender
+            .send_blocking(StorvscCompletion { completion: None })
+            .unwrap();
+    }
+}
+
+/// Errors resulting from storvsc.
+#[derive(Debug, Error)]
+pub enum StorvscError {
+    /// Packet error.
+    #[error("packet error")]
+    PacketError(#[source] PacketError),
+    /// Queue error.
+    #[error("queue error")]
+    Queue(#[source] queue::Error),
+    /// Queue out of space.
+    #[error("queue should have enough space but no longer does")]
+    NotEnoughSpace,
+    /// Unsupported protocol version.
+    #[error("requested protocol version unsupported by storvsp")]
+    UnsupportedProtocolVersion,
+    /// Unexpected protocol data or operation.
+    #[error("unexpected protocol data or operation")]
+    UnexpectedOperation,
+    /// Error notifying completion of operation.
+    #[error("error notifying completion of operation")]
+    NotificationError,
+    /// Error sending request to storvsc driver.
+    #[error("error sending request to storvsc")]
+    RequestError,
+    /// Error receiving new operations from channel.
+    #[error("error receiving new operations from channel")]
+    RequestReceiveError(#[source] RecvError),
+    /// Error waiting for completion of operation.
+    #[error("error waiting for completion of operation")]
+    CompletionError(#[source] RecvError),
+    /// Operation cancelled.
+    #[error("pending operation cancelled")]
+    Cancelled,
+    /// Storvsc driver not fully initialized.
+    #[error("driver not initialized")]
+    Uninitialized,
+}
+
+/// Errors with packet parsing between storvsc and storvsp.
+#[derive(Debug, Error)]
+pub enum PacketError {
+    /// Not transactional.
+    #[error("Not transactional")]
+    NotTransactional,
+    /// Unenxpected transaction.
+    #[error("Unexpected transaction {0:?}")]
+    UnexpectedTransaction(u64),
+    /// Unexpected status.
+    #[error("Unexpected status {0:?}")]
+    UnexpectedStatus(storvsp_protocol::NtStatus),
+    /// Unrecognzied operation.
+    #[error("Unrecognized operation {0:?}")]
+    UnrecognizedOperation(storvsp_protocol::Operation),
+    /// Invalid packet type.
+    #[error("Invalid packet type")]
+    InvalidPacketType,
+    /// Invalid data transfer length.
+    #[error("Invalid data transfer length")]
+    InvalidDataTransferLength,
+    /// Access error.
+    #[error("Access error: {0}")]
```

Comment:
Don't print inner errors that are also marked via `#[source]`--this will result in the inner error's string being printed twice.